mitodl/mitxpro
# Changed username generation to be based on users' full names

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #1079 

#### What's this PR do?
Changes username generation to be based on users' full names.

#### How should this be manually tested?
- Sign up a new user
- Sign up an additional user and use the same Full Name
- Check that the two new users have username values of `<full_name_as_username>` and `<full_name_as_username1>`

To test the cases where an available username is found, but is taken before the pipeline actually creates it, do the following:
- Open a Django shell
- Run the following lines (and leave the shell open):
    ```python
    FULL_NAME = "<your_full_name_from_the_example_above>"
    SUFFIX = 2
    from users.utils import usernameify
    from users.factories import UserFactory
    from django.contrib.auth import get_user_model
    User = get_user_model()
    def get_or_create_test_user(username_base, suffix):
        return User.objects.get_or_create(
            username="{}{}".format(username_base, str(suffix)),
            email="{}@example.com".format(username_base),
            defaults=dict(**{"name": UserFactory.build().name})
        )
    
    ### PREPARE THIS LINE BUT DO NOT RUN IT YET!
    get_or_create_test_user(, SUFFIX + 1)
    ```
- Add this line as the first line of the `try` clause in `authentication.pipeline.user.create_user_via_email`: `import time; time.sleep(15)`
- Create a new user with the same full name as you used above
- After submitting the registration form that asks for your full name, the web service execution should go to sleep. During that sleep time, run the bottom line of the code block above to take the available username that the pipeline found
- After the sleep finishes, check that your new user has a username with a suffix of `3`, e.g.: `johndoe3` (the suffix will be N + 1 where N = the suffix of the user you created above in django shell)

#### Any background context you want to provide?
First, the user's full name is turned into a username-friendly format, then we attempt to save. If there's a collision, the pipeline code tries to find an available username by adding a numerical suffix (e.g.: `johndoe` -> `johndoe1`). The suffix chosen is 1 greater than the highest suffix that exists in the database (e.g.: if `johndoe9` exists, `johndoe10` will be used). If the number of suffix digits causes the username to exceed the username character limit of 26, characters will be removed from the end of the username to make room (e.g.: `abcdefghijklmnopqrstuvwxyz` -> `abcdefghijklmnopqrstuvwxy1`, `abcdefghijklmnopqrstuvwxy9` -> `abcdefghijklmnopqrstuvwx10`)
